sdks/python/apache_beam/pipeline.py

```diff
@@ -976,6 +981,28 @@ def leave_composite_transform(self, transform_node):
     pass
 
 
+class ResourceHintsPropagator(PipelineVisitor):
+  """Propagates resource hints set on composite transforms to subtransforms.
+  """
+  def visit_transform(self, transform_node):
+    # type: (AppliedPTransform) -> None
+    if (transform_node.parent is not None and
+        transform_node.parent.transform is not None and
+        transform_node.parent.transform.get_resource_hints()):
+      transform_node.transform._merge_hints_from_outer_composite(
```

Comment:
Looking over this code (here and elsewhere), it seems preferable for resource hints to be attached to the transform_node rather than the transform. We can populate the hints on the node when constructing it from a transform. This also solves the issue that a single Transform can be applied in multiple places, gives us a place to put the "outermost" ones that come from the pipeline options, and allows users to override `get_resource_hints` with the expected behavior. (It also solves the issue that PTransforms are allowed to be re-used, but their application in the pipeline shouldn't be.)

Reply:
Done

Reply:
Good catch re: multiple applications of the same transform, added a test.